### ISSUE TITLE
Distinguish meaning of fields, e.g. straight alpha and pre-multiplied alpha

### ISSUE THREAD

**Issue by goretkin** [JuliaGraphics/ColorTypes.jl#167]
The example in https://github.com/JuliaGraphics/ColorTypes.jl/issues/166 made me wonder whether `ColorTypes.jl` should try to encode the difference between e.g. pre-multiplied alpha and straight alpha. This kind of semantic information would be necessary to prevent the sort of Hungarian method notation below when trying to implement an alpha compositing operation.

 (pma and sa for premultiplied alpha and straight alpha):

```julia
function over_pma(top::TransparentColor, bot::TransparentColor)
  out_T = promote_type(typeof(bot), typeof(top))
  @show out_T
  top_α = alpha(top)
  bot_α = alpha(bot) * (1 - top_α)
  # assume RGB is premultiplied alpha
  out_α = top_α + bot_α
  out_c = RGB(top) + RGB(bot) * (1 - top_α)
  out_T(out_c, out_α)
end

over_sa(top::TransparentColor, bot::TransparentColor) = over_pma(sa_to_pma(top), sa_to_pma(bot))
```

This could be taken to an extreme, where there is annotation to know whether a color represents, I don't know, transmissivity or emission. I think the two types of interpretation of color + alpha is a more practical matter, as different APIs use different formats: https://microsoft.github.io/Win2D/html/PremultipliedAlpha.htm .

**Comment by kimikage** [comment]
Whether the difference should be handled in the "type" or in the "function" (method) is controversial.
IMO, "ideally", we should create new abstract types like `PMAColor` and `PMColorA` under `TransparentColor`.
However, I doubt that "concrete" types need to be defined in ColorTypes.jl. In particular, defining the constructor is an issue.

The other thing we have to consider is whether alpha compositing is the responsibility of Colors.jl.

Also, be careful what `colorant"transparent"` should return. In [CSS Color Level 3](https://www.w3.org/TR/css-color-3/#transparent), it means a transparent "black" (i.e. `colorant"rgba(0,0,0,0)"`). This can be a problem with gradations, and in fact, different web browsers behave differently. PMA can help resolve the gradation problem, but it should affect many packages.

**Comment by kimikage** [comment]
In addition, the PMA is not fully compatible with cylindrical spaces, i.e. color types with hue (e.g. HSV, LCHab). Although there is no need to prohibit users from using such cylindrical PMA colors, we need to consider what ColorTypes.jl should provide.

Probably some of the methods currently used assume that `TransparentColor` can be divided into `AlphaColor` and `ColorAlpha`. Therefore, it is dangerous to add the PMA types directly under `TransparentColor`.
Adding `PMAColor` under `AlphaColor` will have less impact on other packages, but will break the balance of type hierarchy. Of course, adding `SAColor` under `AlphaColor` is a nightmare.

First of all, what about making a PMA type that does not belong to the hierarchy? I believe that the package where those types are defined should be ColorTypes.jl, but I don't understand where those types will be used.

**Comment by timholy** [comment]
Unless there's a very clear path to resolving this, I'd be quite hesitant to encode this in the type system. The type system is already complex enough as it is, and it's only just complicated enough to handle the diversity of storage-types you find in the wild. (E.g., BGRA buffers passed back and forth between OpenGL and RAM.)

**Comment by kimikage** [comment]
BTW, which package implements [blend modes](https://en.wikipedia.org/wiki/Blend_modes)?

**Comment by timholy** [comment]
I don't know of one, but that doesn't mean it doesn't exist. When I did some compositing I went through Cairo.

**Comment by goretkin** [comment]
I just came across that TIFF can support both alpha modes:
https://github.com/ianshmean/ImageIO.jl/blob/e7bbf7de0b6b12e58243f0ce54f0167244d84038/gen/libtiff/ref_headers/tiff.h#L306-L307

**Comment by timholy** [comment]
Makes sense, but also a can of worms. I wonder how many images there are out in the wild of each encoding?

**Comment by goretkin** [comment]
Yeah, I wonder too. I don't think it's a very _compelling_ reason, in any case. Just a reason.

**Comment by timholy** [comment]
I guess I favor declaring & documenting that we use straight alpha. Then it becomes the responsibility of IO to convert into that format.

**Comment by kimikage** [comment]
I agree with @timholy.
However, despite the fact that the PMA is a common technique for speeding up and preventing false colors, the conversion to SA (and re-conversion to PMA) is lossy and requires additional costs.
I think that the actual use cases or interfaces should be clarified in order to support the PMA natively in the future, or now.

**Comment by goretkin** [comment]
> I guess I favor declaring & documenting that we use straight alpha. Then it becomes the responsibility of IO to convert into that format.

I'm not sure if you meant in this package itself. But I don't think the package needs to make such a decision. There could be a declaration that it uses types to encode

- color space (family?) (e.g. Lab, RGB, ...)
- Maybe™ an alpha channel (i.e. absence of an alpha channel, or the alpha channel)
- storage order of the previous channels
- explicitly does not encode the meaning of the alpha channel for `TransparentColor`

If it turns out that encoding the alpha type is worthwhile, one design that wouldn't require up-ending everything by introducing a type parameter for every `<:TransparentColor`, or `_PMA`/_`_SA` concrete type for every existing type is something like

```ConcreteColorAlpha{C<:Color, T<:Number, AlphaType}
  color::C
  alpha::T
```
(where something, not the type system, could ensure that the `T<:Number` of `Color` and the `T<:Number` of `ConcreteColorAlpha` match)

but I don't know enough about Julia's struct layouts and `reinterpret` to know just how terrible of an idea it is to nest structs for the purposes of `ColorTypes.jl`

**Comment by timholy** [comment]
> one design that wouldn't require up-ending ...

Sorry, but I do not understand what you're suggesting. Did you omit some code? You just have two type-asserts above.

**Comment by goretkin** [comment]
Oops, I had a markdown error. Just edited. Thanks!

**Comment by timholy** [comment]
That was the original design, but we abandoned it because if `c::RGBA` lots of people wanted to be able to say `c.r` and have it work. BUT, that was before we had trait-functions like `red(c)`, and before we could overload `getproperty`. It might be worth reconsidering that decision now.

**Comment by kimikage** [comment]
I am conservative about breaking changes in the types (hierarchies).
However, if we are going to make the major changes, I think it is better not to distinguish between color components and alpha. My idea is to make the color structs tuple-like objects of color components or alphas. That simplifies and flattens the hierarchy.

**Comment by timholy** [comment]
Yeah, even with the improvements to Julia, I'm not at all sure the nested design really will be better. At the moment we achieve "type-nesting" via the abstract type hierarchy. It would be an *awful* lot of work to switch now, and if it's just for this one issue then I think it's certainly not worth it. If there were other benefits then it would be more worthy of consideration.

**Comment by kimikage** [comment]
cf. https://github.com/JuliaGraphics/ColorVectorSpace.jl/issues/126#issuecomment-601980322

[ColorBlendModes.jl](https://github.com/kimikage/ColorBlendModes.jl) is a just new package, but I think it is possible to support and "grow" the pre-multiplied alpha colors there in the future. After the features of the pre-multiplied alpha colors have matured, it is a good idea to move their types to ColorTypes.jl. Of course, there are no concrete plans at this time.
